File: algorithms/edmonds-blossom-simple/cpp/edmonds_blossom_simple.cpp
/*
 * Edmonds' Blossom Algorithm (Simple Version) for Maximum Cardinality Matching
 * Time complexity: O(V⁴)
 * 
 * This is the straightforward implementation that finds one augmenting path per iteration.
 */

#include <iostream>
#include <fstream>
#include <vector>
#include <unordered_map>
#include <unordered_set>
#include <queue>
#include <string>
#include <chrono>
#include <cstdlib>
#include <ctime>

class EdmondsBlossomSimple {
private:
    std::unordered_set<std::string> vertices;
    std::unordered_map<std::string, std::unordered_set<std::string>> graph;
    std::unordered_map<std::string, std::string> mate;
    
    // For finding augmenting paths
    std::unordered_map<std::string, std::string> parent;
    std::unordered_map<std::string, std::string> base;
    std::unordered_map<std::string, std::string> in_blossom;
    
    bool is_outer(const std::string& v) {
        if (parent.find(v) == parent.end()) {
            return false;
        }
        
        std::string current = v;
        int distance = 0;
        std::unordered_set<std::string> visited;
        
        while (current != parent[current] && visited.find(current) == visited.end()) {
            visited.insert(current);
            current = parent[current];
            distance++;
        }
        
        return distance % 2 == 0;
    }
    
    std::string find_blossom(const std::string& v, const std::string& w) {
        // Mark all ancestors of v
        std::unordered_set<std::string> path_v;
        std::string current = v;
        
        while (parent.find(current) != parent.end()) {
            path_v.insert(base[current]);
            if (current == parent[current]) break;
            current = parent[current];
        }
        
        // Find first common ancestor from w
        current = w;
        while (parent.find(current) != parent.end()) {
            if (path_v.find(base[current]) != path_v.end()) {
                return base[current];
            }
            if (current == parent[current]) break;
            current = parent[current];
        }
        
        return v; // Fallback
    }
    
    void contract_blossom(const std::string& blossom_base, const std::string& v, const std::string& w) {
        // Path from v to base
        std::string current = v;
        while (base[current] != blossom_base) {
            base[current] = blossom_base;
            if (!mate[current].empty()) {
                base[mate[current]] = blossom_base;
                current = parent[mate[current]];
            } else {
                break;
            }
        }
        
        // Path from w to base
        current = w;
        while (base[current] != blossom_base) {
            base[current] = blossom_base;
            if (!mate[current].empty()) {
                base[mate[current]] = blossom_base;
                current = parent[mate[current]];
            } else {
                break;
            }
        }
        
        base[v] = blossom_base;
        base[w] = blossom_base;
    }
    
    std::vector<std::string> build_path(const std::string& v, const std::string& w) {
        std::vector<std::string> path = {w, v};
        
        std::string current = v;
        while (parent.find(current) != parent.end() && parent[current] != current) {
            std::string p = parent[current];
            path.push_back(p);
            current = p;
        }
        
        return path;
    }
    
    void augment(const std::vector<std::string>& path) {
        for (size_t i = 0; i + 1 < path.size(); i += 2) {
            mate[path[i]] = path[i + 1];
            mate[path[i + 1]] = path[i];
        }
    }

public:
    EdmondsBlossomSimple(const std::vector<std::string>& vertex_list,
                         const std::vector<std::pair<std::string, std::string>>& edges) {
        
        vertices.insert(vertex_list.begin(), vertex_list.end());
        
        // Build adjacency list (undirected graph)
        for (const auto& edge : edges) {
            if (vertices.count(edge.first) && vertices.count(edge.second) && edge.first != edge.second) {
                graph[edge.first].insert(edge.second);
                graph[edge.second].insert(edge.first);
            }
        }
        
        // Initialize matching
        for (const auto& v : vertices) {
            mate[v] = "";
        }
    }
    
    std::vector<std::string> find_augmenting_path(const std::string& start) {
        // Initialize structures
        parent.clear();
        base.clear();
        in_blossom.clear();
        
        parent[start] = start;
        for (const auto& v : vertices) {
            base[v] = v;
            in_blossom[v] = v;
        }
        
        std::queue<std::string> q;
        q.push(start);
        
        while (!q.empty()) {
            std::string v = q.front();
            q.pop();
            
            for (const auto& w : graph[v]) {
                // Skip if already in tree at same level
                if (base[v] == base[w]) {
                    continue;
                }
                
                // If w is unmatched, we found an augmenting path!
                if (mate[w].empty()) {
                    return build_path(v, w);
                }
                
                // If w is not in tree yet
                if (parent.find(w) == parent.end()) {
                    parent[w] = v;
                    std::string mate_w = mate[w];
                    if (!mate_w.empty()) {
                        parent[mate_w] = w;
                        q.push(mate_w);
                    }
                }
                // If w is in tree at even distance (blossom found)
                else if (is_outer(w)) {
                    std::string blossom_base = find_blossom(v, w);
                    contract_blossom(blossom_base, v, w);
                    q.push(blossom_base);
                }
            }
        }
        
        return std::vector<std::string>();
    }
    
    std::vector<std::pair<std::string, std::string>> maximum_matching() {
        bool improved = true;
        
        while (improved) {
            improved = false;
            
            for (const auto& v : vertices) {
                if (mate[v].empty()) {
                    std::vector<std::string> path = find_augmenting_path(v);
                    if (!path.empty()) {
                        augment(path);
                        improved = true;
                        break;
                    }
                }
            }
        }
        
        // Build the matching set
        std::vector<std::pair<std::string, std::string>> matching;
        std::unordered_set<std::string> seen;
        
        for (const auto& u : vertices) {
            std::string v = mate[u];
            if (!v.empty() && seen.find(v) == seen.end()) {
                matching.push_back({u, v});
                seen.insert(u);
            }
        }
        
        return matching;
    }
};

// Load graph from file
bool load_graph_from_file(const std::string& filename,
                          std::vector<std::string>& vertices,
                          std::vector<std::pair<std::string, std::string>>& edges) {
    std::ifstream file(filename);
    if (!file.is_open()) {
        return false;
    }
    
    int vertex_count, edge_count;
    file >> vertex_count >> edge_count;
    
    vertices.clear();
    edges.clear();
    
    for (int i = 0; i < vertex_count; i++) {
        vertices.push_back("V" + std::to_string(i));
    }
    
    for (int i = 0; i < edge_count; i++) {
        int u, v;
        file >> u >> v;
        edges.push_back({"V" + std::to_string(u), "V" + std::to_string(v)});
    }
    
    file.close();
    return true;
}

// Generate random test graph
std::vector<std::pair<std::string, std::string>> generate_test_graph(
    int n, double edge_probability, std::vector<std::string>& vertices) {
    
    vertices.clear();
    for (int i = 0; i < n; i++) {
        vertices.push_back("V" + std::to_string(i));
    }
    
    std::vector<std::pair<std::string, std::string>> edges;
    for (int i = 0; i < n; i++) {
        for (int j = i + 1; j < n; j++) {
            if ((double)rand() / RAND_MAX < edge_probability) {
                edges.push_back({vertices[i], vertices[j]});
            }
        }
    }
    
    return edges;
}

void run_example(const std::vector<std::string>& vertices,
                 const std::vector<std::pair<std::string, std::string>>& edges,
                 const std::string& description) {
    std::cout << description << "\n";
    std::cout << "Graph: " << vertices.size() << " vertices, " << edges.size() << " edges\n";
    
    auto start = std::chrono::high_resolution_clock::now();
    EdmondsBlossomSimple eb(vertices, edges);
    auto matching = eb.maximum_matching();
    auto end = std::chrono::high_resolution_clock::now();
    
    auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(end - start);
    
    std::cout << "Matching size: " << matching.size() << "\n";
    if (matching.size() <= 10) {
        std::cout << "Matching: ";
        for (const auto& edge : matching) {
            std::cout << "(" << edge.first << "," << edge.second << ") ";
        }
        std::cout << "\n";
    }
    std::cout << "Execution time: " << duration.count() << " ms\n";
    std::cout << "\n";
}

int main(int argc, char* argv[]) {
    std::cout << "Edmonds' Blossom Algorithm (Simple) - C++ Implementation\n";
    std::cout << "========================================================\n\n";
    
    srand(time(0));
    
    // Check if a file was provided
    if (argc > 1) {
        std::string filename = argv[1];
        std::cout << "Loading graph from: " << filename << "\n";
        
        std::vector<std::string> vertices;
        std::vector<std::pair<std::string, std::string>> edges;
        
        if (!load_graph_from_file(filename, vertices, edges)) {
            std::cerr << "Error: Could not open file '" << filename << "'\n";
            return 1;
        }
        
        run_example(vertices, edges, "File: " + filename);
    } else {
        // Run built-in examples
        std::cout << "Running built-in examples (use: ./edmonds_blossom_simple_cpp <filename> to load from file)\n\n";
        
        // Example 1: Simple graph with triangle
        std::vector<std::string> vertices = {"A", "B", "C", "D", "E"};
        std::vector<std::pair<std::string, std::string>> edges = {
            {"A", "B"}, {"B", "C"}, {"C", "A"},  // Triangle
            {"C", "D"}, {"D", "E"}
        };
        run_example(vertices, edges, "Example 1: Graph with triangle (blossom)");
        
        // Example 2: Larger random graph
        vertices.clear();
        edges = generate_test_graph(20, 0.2, vertices);
        run_example(vertices, edges, "Example 2: Random graph (20 vertices)");
        
        // Benchmark
        std::cout << "Benchmarking with larger graph...\n";
        vertices.clear();
        edges = generate_test_graph(100, 0.1, vertices);
        run_example(vertices, edges, "Benchmark: Random graph (100 vertices)");
    }
    
    return 0;
}

File: algorithms/edmonds-blossom-simple/rust/edmonds_blossom_simple.rs
/*
 * Edmonds' Blossom Algorithm (Simple Version) for Maximum Cardinality Matching
 * Time complexity: O(V⁴)
 * 
 * This is the straightforward implementation that finds one augmenting path per iteration.
 */

use std::collections::{HashMap, HashSet, VecDeque};
use std::env;
use std::fs::File;
use std::io::{self, BufRead, BufReader};
use std::time::Instant;

struct EdmondsBlossomSimple {
    vertices: HashSet<String>,
    graph: HashMap<String, HashSet<String>>,
    mate: HashMap<String, Option<String>>,
}

impl EdmondsBlossomSimple {
    fn new(vertex_list: &[String], edges: &[(String, String)]) -> Self {
        let vertices: HashSet<String> = vertex_list.iter().cloned().collect();
        let mut graph: HashMap<String, HashSet<String>> = HashMap::new();
        
        // Build adjacency list (undirected graph)
        for (u, v) in edges {
            if vertices.contains(u) && vertices.contains(v) && u != v {
                graph.entry(u.clone()).or_insert_with(HashSet::new).insert(v.clone());
                graph.entry(v.clone()).or_insert_with(HashSet::new).insert(u.clone());
            }
        }
        
        // Initialize matching
        let mut mate = HashMap::new();
        for v in &vertices {
            mate.insert(v.clone(), None);
        }
        
        EdmondsBlossomSimple {
            vertices,
            graph,
            mate,
        }
    }
    
    fn is_outer(&self, v: &str, parent: &HashMap<String, String>) -> bool {
        if !parent.contains_key(v) {
            return false;
        }
        
        let mut current = v.to_string();
        let mut distance = 0;
        let mut visited = HashSet::new();
        
        while let Some(p) = parent.get(&current) {
            if &current == p || visited.contains(&current) {
                break;
            }
            visited.insert(current.clone());
            current = p.clone();
            distance += 1;
        }
        
        distance % 2 == 0
    }
    
    fn find_blossom(&self, v: &str, w: &str, 
                    parent: &HashMap<String, String>,
                    base: &HashMap<String, String>) -> String {
        // Mark all ancestors of v
        let mut path_v = HashSet::new();
        let mut current = v.to_string();
        
        while let Some(p) = parent.get(&current) {
            path_v.insert(base.get(&current).unwrap_or(&current).clone());
            if &current == p {
                break;
            }
            current = p.clone();
        }
        
        // Find first common ancestor from w
        current = w.to_string();
        while let Some(p) = parent.get(&current) {
            let base_current = base.get(&current).unwrap_or(&current);
            if path_v.contains(base_current) {
                return base_current.clone();
            }
            if &current == p {
                break;
            }
            current = p.clone();
        }
        
        v.to_string() // Fallback
    }
    
    fn contract_blossom(&self, blossom_base: &str, v: &str, w: &str,
                       base: &mut HashMap<String, String>,
                       parent: &HashMap<String, String>) {
        // Path from v to base
        let mut current = v.to_string();
        while base.get(&current).unwrap_or(&current) != blossom_base {
            base.insert(current.clone(), blossom_base.to_string());
            
            if let Some(Some(mate_current)) = self.mate.get(&current) {
                base.insert(mate_current.clone(), blossom_base.to_string());
                if let Some(next) = parent.get(mate_current) {
                    current = next.clone();
                } else {
                    break;
                }
            } else {
                break;
            }
        }
        
        // Path from w to base
        current = w.to_string();
        while base.get(&current).unwrap_or(&current) != blossom_base {
            base.insert(current.clone(), blossom_base.to_string());
            
            if let Some(Some(mate_current)) = self.mate.get(&current) {
                base.insert(mate_current.clone(), blossom_base.to_string());
                if let Some(next) = parent.get(mate_current) {
                    current = next.clone();
                } else {
                    break;
                }
            } else {
                break;
            }
        }
        
        base.insert(v.to_string(), blossom_base.to_string());
        base.insert(w.to_string(), blossom_base.to_string());
    }
    
    fn build_path(&self, v: &str, w: &str, parent: &HashMap<String, String>) -> Vec<String> {
        let mut path = vec![w.to_string(), v.to_string()];
        
        let mut current = v.to_string();
        while let Some(p) = parent.get(&current) {
            if &current == p {
                break;
            }
            path.push(p.clone());
            current = p.clone();
        }
        
        path
    }
    
    fn augment(&mut self, path: &[String]) {
        for i in (0..path.len().saturating_sub(1)).step_by(2) {
            let u = &path[i];
            let v = &path[i + 1];
            self.mate.insert(u.clone(), Some(v.clone()));
            self.mate.insert(v.clone(), Some(u.clone()));
        }
    }
    
    fn find_augmenting_path(&self, start: &str) -> Vec<String> {
        let mut parent: HashMap<String, String> = HashMap::new();
        let mut base: HashMap<String, String> = HashMap::new();
        
        parent.insert(start.to_string(), start.to_string());
        for v in &self.vertices {
            base.insert(v.clone(), v.clone());
        }
        
        let mut queue = VecDeque::new();
        queue.push_back(start.to_string());
        
        while let Some(v) = queue.pop_front() {
            if let Some(neighbors) = self.graph.get(&v) {
                for w in neighbors {
                    let base_v = base.get(&v).unwrap_or(&v);
                    let base_w = base.get(w).unwrap_or(w);
                    
                    // Skip if already in tree at same level
                    if base_v == base_w {
                        continue;
                    }
                    
                    // If w is unmatched, we found an augmenting path!
                    if self.mate.get(w).unwrap_or(&None).is_none() {
                        return self.build_path(&v, w, &parent);
                    }
                    
                    // If w is not in tree yet
                    if !parent.contains_key(w) {
                        parent.insert(w.clone(), v.clone());
                        if let Some(Some(mate_w)) = self.mate.get(w) {
                            parent.insert(mate_w.clone(), w.clone());
                            queue.push_back(mate_w.clone());
                        }
                    }
                    // If w is in tree at even distance (blossom found)
                    else if self.is_outer(w, &parent) {
                        let blossom_base = self.find_blossom(&v, w, &parent, &base);
                        self.contract_blossom(&blossom_base, &v, w, &mut base, &parent);
                        queue.push_back(blossom_base);
                    }
                }
            }
        }
        
        Vec::new()
    }
    
    fn maximum_matching(&mut self) -> Vec<(String, String)> {
        let mut improved = true;
        
        while improved {
            improved = false;
            
            let vertices: Vec<String> = self.vertices.iter().cloned().collect();
            for v in vertices {
                if self.mate.get(&v).unwrap_or(&None).is_none() {
                    let path = self.find_augmenting_path(&v);
                    if !path.is_empty() {
                        self.augment(&path);
                        improved = true;
                        break;
                    }
                }
            }
        }
        
        // Build the matching set
        let mut matching = Vec::new();
        let mut seen = HashSet::new();
        
        for u in &self.vertices {
            if let Some(Some(v)) = self.mate.get(u) {
                if !seen.contains(v) {
                    matching.push((u.clone(), v.clone()));
                    seen.insert(u.clone());
                }
            }
        }
        
        matching
    }
}

// Load graph from file
fn load_graph_from_file(filename: &str) -> io::Result<(Vec<String>, Vec<(String, String)>)> {
    let file = File::open(filename)?;
    let mut lines = BufReader::new(file).lines();
    
    let first_line = lines.next().ok_or(io::Error::new(
        io::ErrorKind::InvalidData,
        "Empty file"
    ))??;
    
    let parts: Vec<usize> = first_line
        .split_whitespace()
        .map(|s| s.parse().map_err(|_| io::Error::new(
            io::ErrorKind::InvalidData,
            "Invalid number format"
        )))
        .collect::<Result<Vec<_>, _>>()?;
    
    if parts.len() != 2 {
        return Err(io::Error::new(
            io::ErrorKind::InvalidData,
            "First line must have 2 numbers"
        ));
    }
    
    let (vertex_count, edge_count) = (parts[0], parts[1]);
    
    let vertices: Vec<String> = (0..vertex_count).map(|i| format!("V{}", i)).collect();
    
    let mut edges = Vec::new();
    for line in lines.take(edge_count) {
        let line = line?;
        let nums: Vec<usize> = line
            .split_whitespace()
            .map(|s| s.parse().map_err(|_| io::Error::new(
                io::ErrorKind::InvalidData,
                "Invalid edge format"
            )))
            .collect::<Result<Vec<_>, _>>()?;
        
        if nums.len() != 2 {
            return Err(io::Error::new(
                io::ErrorKind::InvalidData,
                "Each edge line must have 2 numbers"
            ));
        }
        
        edges.push((format!("V{}", nums[0]), format!("V{}", nums[1])));
    }
    
    Ok((vertices, edges))
}

// Generate deterministic test graph
fn generate_test_graph(n: usize, edge_probability_percent: usize) -> (Vec<String>, Vec<(String, String)>) {
    let vertices: Vec<String> = (0..n).map(|i| format!("V{}", i)).collect();
    let mut edges = Vec::new();
    
    for i in 0..n {
        for j in (i + 1)..n {
            // Deterministic "random" based on i and j
            if ((i * 17 + j * 31) % 100) < edge_probability_percent {
                edges.push((vertices[i].clone(), vertices[j].clone()));
            }
        }
    }
    
    (vertices, edges)
}

fn run_example(vertices: &[String], edges: &[(String, String)], description: &str) {
    println!("{}", description);
    println!("Graph: {} vertices, {} edges", vertices.len(), edges.len());
    
    let start = Instant::now();
    let mut eb = EdmondsBlossomSimple::new(vertices, edges);
    let matching = eb.maximum_matching();
    let duration = start.elapsed();
    
    println!("Matching size: {}", matching.len());
    if matching.len() <= 10 {
        print!("Matching: ");
        for (u, v) in &matching {
            print!("({},{}) ", u, v);
        }
        println!();
    }
    println!("Execution time: {} ms", duration.as_millis());
    println!();
}

fn main() {
    println!("Edmonds' Blossom Algorithm (Simple) - Rust Implementation");
    println!("==========================================================\n");
    
    let args: Vec<String> = env::args().collect();
    
    // Check if a file was provided
    if args.len() > 1 {
        let filename = &args[1];
        println!("Loading graph from: {}", filename);
        
        match load_graph_from_file(filename) {
            Ok((vertices, edges)) => {
                run_example(&vertices, &edges, &format!("File: {}", filename));
            }
            Err(e) => {
                eprintln!("Error loading file: {}", e);
                std::process::exit(1);
            }
        }
    } else {
        // Run built-in examples
        println!("Running built-in examples (use: ./edmonds_blossom_simple_rust <filename> to load from file)\n");
        
        // Example 1: Simple graph with triangle
        let vertices: Vec<String> = vec!["A", "B", "C", "D", "E"]
            .into_iter()
            .map(|s| s.to_string())
            .collect();
        let edges: Vec<(String, String)> = vec![
            ("A", "B"), ("B", "C"), ("C", "A"),  // Triangle
            ("C", "D"), ("D", "E"),
        ]
        .into_iter()
        .map(|(a, b)| (a.to_string(), b.to_string()))
        .collect();
        
        run_example(&vertices, &edges, "Example 1: Graph with triangle (blossom)");
        
        // Example 2: Larger random graph
        let (vertices, edges) = generate_test_graph(20, 20);
        run_example(&vertices, &edges, "Example 2: Random graph (20 vertices)");
        
        // Benchmark
        println!("Benchmarking with larger graph...");
        let (vertices, edges) = generate_test_graph(100, 10);
        run_example(&vertices, &edges, "Benchmark: Random graph (100 vertices)");
    }
}
